Added support for writing received DICOM content to Writable streams

PantelisGeorgiadis opened the pull request:
An attempt was made to implement C-STORE SCP streaming support, based on the ideas discussed in https://github.com/PantelisGeorgiadis/dcmjs-dimse/issues/39. 
More specifically, the `Scp` class was enriched with two extra functions that give the caller the oppurtunity to create the stream that will receive the incoming fragment buffers and to translate the accumulated fragments back into a `Dataset`.
The functions are the `createStoreWritableStream` and the `createDatasetFromStoreWritableStream`, respectively. The first one provides to the user the accepted presentation context (to assess the dataset type and transfer syntax) and expects a Writable stream to be returned. The latter provides to the user the previously created Writable stream, the accepted presentation context and callback that could be used to return the parsed `Dataset`. The default implementation still uses memory buffers.

Bellow is a sample implementation of an `Scp` class that accumulates the incoming C-STORE data to temp file streams (using the `temp` library). Once the reception is over (last PDV fragment received), the files are read back into a `Dataset` which are passed to the `Scp.cStoreRequest` method for further processing (this is not a mandatory step - the callback can return undefined if there is no interest for passing the dataset in the `Scp.cStoreRequest` method).
```js
const temp = require('temp');

class StreamingScp extends Scp {
  constructor(socket, opts) {
    super(socket, opts);
    this.association = undefined;

    temp.track();
  }

  createStoreWritableStream(acceptedPresentationContext) {
    return temp.createWriteStream();
  }

  createDatasetFromStoreWritableStream(writable, acceptedPresentationContext, callback) {
    writable.on('finish', () => {
      const path = writable.path;
      const datasetBuffer = fs.readFileSync(path);
      const dataset = new Dataset(
        datasetBuffer,
        acceptedPresentationContext.getAcceptedTransferSyntaxUid()
      );
      callback(dataset);
    });
    writable.on('error', (err) => {
      callback(undefined);
    });
  }

  associationRequested(association) {
    this.association = association;

    const contexts = association.getPresentationContexts();
    contexts.forEach((c) => {
      const context = association.getPresentationContext(c.id);
      if (Object.values(StorageClass).includes(context.getAbstractSyntaxUid())) {
        const transferSyntaxes = context.getTransferSyntaxUids();
        transferSyntaxes.forEach((transferSyntax) => {
          if (
            transferSyntax === TransferSyntax.ImplicitVRLittleEndian ||
            transferSyntax === TransferSyntax.ExplicitVRLittleEndian
          ) {
            context.setResult(PresentationContextResult.Accept, transferSyntax);
          } else {
            context.setResult(PresentationContextResult.RejectTransferSyntaxesNotSupported);
          }
        });
      } else {
        context.setResult(PresentationContextResult.RejectAbstractSyntaxNotSupported);
      }
    });
    this.sendAssociationAccept();
  }

  cStoreRequest(request, callback) {
    console.log(request.getDataset());

    const response = CStoreResponse.fromRequest(request);
    response.setStatus(Status.Success);
    callback(response);
  }

  associationReleaseRequested() {
    this.sendAssociationReleaseResponse();
  }
}
```

@richard-viney, @kinsonho, please review and let me know of your thoughts.

kinsonho commented on src/Network.js:
```diff
@@ -480,18 +512,37 @@ class Network extends AsyncEventEmitter {
+        }
+
+        const stream = this.dimseStream || this.dimseStoreStream;
+        stream.write(pdv.getValue());
```
Here it should check the return value of write(). If it is not true, then we need to apply backpressure and until the drain event to resume. https://nodejs.org/docs/latest/api/stream.html#event-drain

PantelisGeorgiadis replied:
Nice catch! Thank you Kinson! The only thing that bothers me is how to wait in here for the `drain` event... any ideas?

PantelisGeorgiadis replied:
Any ideas on this @kinsonho?

kinsonho replied:
@PantelisGeorgiadis sorry for the late reply.

A pattern that I saw others doing is the following:

    if(!stream.write(pdv.getValue())) {
        await new Promise(resolve => stream.once('drain', resolve));
    }

PantelisGeorgiadis replied:
Thank you @kinsonho! The async/await pattern is not used in this library. Do you see any drawbacks in start using it (e.g. compatibility, etc.)?

kinsonho commented:
With createStoreWriteableStream(), the result is just the dataset, not a full P10 object, right?

For createDatasetFromStoreWriteableStream(), if I would like to have a filtered dataset to subsequent processing after the cStoreRequest, then I will need to split the StoreWriteableStream and create another accumulator writer stream to collect the attributes that I need. Then pass to the createDatasetFromStoreWritableStream(). Correct?

kinsonho commented:
Thanks for working on this. Just some quick comments when reviewing the changes for now. I will try it out later.

PantelisGeorgiadis commented:
@kinsonho, @richard-viney did you have the chance to test this? Any feedback?

kinsonho commented:
Sorry for the long delay. I am planning to work on it this week. I will let
you know.

On Sun, May 28, 2023 at 3:40 PM Pantelis Georgiadis <
***@***.***> wrote:

> @kinsonho <https://github.com/kinsonho>, @richard-viney
> <https://github.com/richard-viney> did you have the chance to test this?
> Any feedback?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PantelisGeorgiadis/dcmjs-dimse/pull/41#issuecomment-1566241466>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3KBIDMFKNOHYOYXVERZV3XIOSZXANCNFSM6AAAAAAV7S36HY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

PantelisGeorgiadis commented:
@richard-viney I just rebased to latest master. My only concern is that we still haven't handled Kinson's comment regarding the stream `drain` event. Looking forward for your test results!

richard-viney commented:
Yes the drain comment is correct. In practice if streaming is going to disk then the incoming network data will most likely be at a slower pace than local disk writes, so you'd get away with it, but for other use cases it could become an issue.

I haven't yet looked at the implications for the library/protocol of adding a wait on the stream here.

richard-viney commented:
Would it be possible to pass the `CStoreRequest` to `createStoreWritableStream()`? This would allow it access to information such as `getAffectedSopInstanceUid()` which would be useful for trying to write the file meta information prior to the main data set coming in, which is what I'm aiming to do.

richard-viney commented:
I've got this working, ended up writing a custom DICOM P10 header prior to streaming the rest to a temporary file, which results in a valid DICOM P10. Still more testing to be done though.

My previous comment about making `createStoreWritableStream` asynchronous isn't important as I worked around it and deferred error handling to the existing `cStoreRequest()` method, which I think is reasonable given that it decides on the response to send to the SCU.

IMO the most important change to add to this branch is making the C-STORE request available, which I did in a fork here: https://github.com/HeartLab/dcmjs-dimse/commit/4192ec93237d0cd7b23f89a063c21a1426f98c03.

The only other issue was regarding `drain()` and back pressure, however this change is still a big improvement even without that issue being addressed. Have you had a chance to look into it? If not, I'll likely do so in the next couple of days.

Thanks.

PantelisGeorgiadis commented:
Great news @richard-viney! Thanks for working on this! I incorporated your changes to the working branch and made an unsuccessful effort to handle stream backpressure, which I reverted… ☹
However, I’m pretty sure that we will find a way to handle it!
